Web Inspector: Improve DataGridItem performance when adding a large number of items to a data grid at once

patrickangle opened the pull request:
#### 8f75ae36260cdc7aff28822500d04e423e1b2c73
<pre>
Web Inspector: Improve DataGridItem performance when adding a large number of items to a data grid at once
<a href="https://bugs.webkit.org/show_bug.cgi?id=242729">https://bugs.webkit.org/show_bug.cgi?id=242729</a>
rdar://96892753

Reviewed by Devin Rousso.

When adding a data grid item to a DataGrid, it is very likely the previous item is closer to the end than the beginning
of the collection of rows, so we should use `lastIndexOf` instead of `indexOf` to find the index of that item, otherwise
for grids with tens of thousands of items (like a timeline recording, for example) we will have to traverse down every
row for each item we add, even though we are adding the items sequentially in bulk. For a tested very large timeline
recording with ~180,000 Layout/Rendering records, this improved view presentation time from ~10 seconds locally to under
a second.

To preserve the performance of inserting an item at the start of the list, we check if the previous item is null, and
skip checking the entire array of rows.

* Source/WebInspectorUI/UserInterface/Views/DataGridNode.js:
(WI.DataGridNode.prototype._attach):

Canonical link: <a href="https://commits.webkit.org/252529@main">https://commits.webkit.org/252529@main</a>
</pre>

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DataGridNode.js:
```diff
@@ -674,7 +674,7 @@ WI.DataGridNode = class DataGridNode extends WI.Object
         if (!this.isPlaceholderNode) {
             var previousGridNode = this.traversePreviousNode(true, true);
-            insertionIndex = this.dataGrid._rows.indexOf(previousGridNode);
+            insertionIndex = this.dataGrid._rows.lastIndexOf(previousGridNode);
```
ha! clever :)

there's probably a bunch of other spots we could do this too.  i do wonder about the cost of this in other situations where we do insert near the front of the list tho :/

maybe we could add some hints somewhere that suggest "we're close to the front" or somethin

patrickangle replied:
We could certainly optimize the null path, which would cover first-item insertion.

JosephPecoraro replied:
I wonder if sub-calls to `_attach` could be optimized to just have the known index. Like adding the children in the `this.expanded` case (line: 693). But I don't recall if sort order could impact things. I also don't recall if `expanded` matters much in the performance heavy cases (like Timeline data grids).

patrickangle committed:
Web Inspector: Improve DataGridItem performance when adding a large number of items to a data grid at once
https://bugs.webkit.org/show_bug.cgi?id=242729
rdar://96892753

Reviewed by Devin Rousso.

When adding a data grid item to a DataGrid, it is very likely the previous item is closer to the end than the beginning
of the collection of rows, so we should use `lastIndexOf` instead of `indexOf` to find the index of that item, otherwise
for grids with tens of thousands of items (like a timeline recording, for example) we will have to traverse down every
row for each item we add, even though we are adding the items sequentially in bulk. For a tested very large timeline
recording with ~180,000 Layout/Rendering records, this improved view presentation time from ~10 seconds locally to under
a second.

To preserve the performance of inserting an item at the start of the list, we check if the previous item is null, and
skip checking the entire array of rows.

* Source/WebInspectorUI/UserInterface/Views/DataGridNode.js:
(WI.DataGridNode.prototype._attach):

Canonical link: https://commits.webkit.org/252529@main
